### ISSUE TITLE
DataFrame P2P Shuffling with stable ordering

### ISSUE THREAD

**Issue by fjetter** [dask/distributed#8421]
The P2P algorithm as is currently does not strictly guarantee ordering. This can be problematic for some order sensitive operations like groupby + first (https://github.com/dask/dask/issues/10034) or for a `drop_duplicates` with `keep` (https://github.com/dask/dask/issues/10708)

It's a little work but should be possible to get P2P to be stable

**Comment by hendrikmakait** [comment]
For the sake of documenting and being precise: When talking about _stable_ ordering, the only thing we can guarantee with P2P is stable ordering between rows of the same shuffle key (i.e., the combination of values of the rows/index we shuffle on). With shuffling as a hashing-based operation, any ordering between keys is impossible.

**Comment by hendrikmakait** [comment]
## A/B test results

### TL;DR

`test_q6 shows a significant increase in peak (and avg.) memory. It's hard to tell how bad this is in reality given the ~10s duration of the query. I'm inclined to move forward with this PR.

### Runtime
<img width="1670" alt="Screenshot 2024-01-12 at 09 44 38" src="https://github.com/dask/distributed/assets/2699097/0504f9dc-b646-418e-b811-f6613036d8c5">

This looks like noise, `test_set_index_on_uber_lyft` got messed up by spot interruption.

### Average Memory
<img width="1670" alt="Screenshot 2024-01-12 at 09 45 03" src="https://github.com/dask/distributed/assets/2699097/dd61102a-b85b-487f-94e2-276db8926c5d">

Mostly noise, `test_q6` looks significant (more on that later).

### Peak Memory

<img width="1670" alt="Screenshot 2024-01-12 at 09 45 11" src="https://github.com/dask/distributed/assets/2699097/f00045cf-4059-4003-976f-c540716c795f">

Once again noise with `test_q6` standing out.


### test_q6

<img width="842" alt="Screenshot 2024-01-12 at 09 45 50" src="https://github.com/dask/distributed/assets/2699097/ef62622b-dfc7-489f-aff9-6b99fc9435f4">
<img width="842" alt="Screenshot 2024-01-12 at 09 45 56" src="https://github.com/dask/distributed/assets/2699097/e3eac236-2cd1-4147-a7bd-fb808ce72b4c">

Looking at the memory plots, it seems as though this PR has a significant impact on peak memory usage. Average memory also increases but remains pretty small.
